docs: Safe Smart Account docs improvements (#653)
safe-global/safe-docs: pages/advanced/smart-account-overview.mdx

import SafeSmartAccount from '../../assets/svg/ic-smart-account.svg'

# How do Safe Smart Accounts work?

## EOAs

[EOAs](/home/glossary#externally-owned-account), for short, are Ethereum accounts that use traditional key pairs. That is, they consist of a single private key that can be used to make transactions and sign messages. If you gain access to that private key, you gain complete control of the account. This means that your private key is a single point of failure between you and your funds being lost.

## Smart accounts

> The other type of Ethereum account is a [smart contract account](/home/glossary#smart-account). Like EOAs, smart contract accounts each have a unique public Ethereum address, and it is impossible to tell them apart from EOAs by looking at an Ethereum address. Smart contract accounts, too, can receive funds and make transactions like EOAs but cannot initiate them. Generally, the key difference is that no single private key is used to verify transactions. Instead, the smart contract code defines the logic behind how the account completes transactions. Smart contracts are programs that run on the Ethereum blockchain and execute when specific conditions are met. Their functionality within contract accounts means that such accounts, in contrast to EOAs, can, for example, implement access rights that specify by whom, how, and under which conditions transactions can be executed, as well as more complex logic.

Reviewer: Let's use "Smart Accounts" in a capitalized way.

# Difference between EOAs and Contract Accounts

![Ethereum Accounts](../../assets/safe-smart-contract-accounts.png)

## Safe Smart Account

> Safe Smart Account is a contract account with multi-signature functionality at its core. It is secure and flexible, and it can be used to manage funds and execute transactions on the Ethereum blockchain.

Reviewer: ... is a Smart Account ...

#### Token callback support

Many new tokens require wallet contracts to implement callbacks. Token standards like **ERC-721** and **ERC-1155** allow contracts to immediately react to receiving tokens through these and make it even possible to reject the transfer altogether.

#### Sponsored transactions

Another core functionality of the Safe is **token payment**. Generally, Ethereum transactions require ETH to pay transaction fees ("gas"). With the Safe, users can pay transaction fees in several supported ERC20 tokens. This is realized via a transaction relay service that accepts those tokens and submits the transactions to the blockchain, paying the gas fee in ETH. With the same functionality, Ether-less transactions can be implemented, where a 3rd party pays transaction fees on behalf of a Safe via the same relay service.

### Architecture

![Safe Smart Accounts Architecture](../../assets/diagram-safe-smart-accounts-architecture.png)

#### Safe Singleton Factory

> The Safe Singleton Factory is a contract that deploys all the Safe related contracts. This contract helps to deploy Safe contracts at the same address across different networks and eventually also helps to deploy Safe proxies at the same address across different networks.

Reviewer: "Safe Smart Account related contracts". (We want to use Safe as a brand, and not to identify Smart Accounts). Also replace "Safe" with "Safe Smart Account" in the next sentence.

##### Guard Management

Guards are used to check if a transaction should be executed or rejected based on the logic defined in the guard. A Guard Manager contract allows you to add, remove, and replace guards. Guards are security critical as a malicious guard could prevent transactions from being executed and block access to funds stored in the Safe. Events are emitted whenever a guard is updated.

#### SafeProxy

A Safe Proxy is a contract that delegates all calls to the Safe Singleton. Deploying a Proxy reduces the cost of creating a Safe account, as the proxy contract's byte code size is less than that of the actual Safe contract code.

![Safe Proxy Creation](../../assets/diagram-safe-smart-account-proxy-creation.png)

### Concepts

#### Owners

Each Safe account maintains its own list of [owners](/home/glossary#owner) in the storage in the form of Ethereum addresses. Owners can be added or removed by other owners.

#### Threshold

> Safe Smart Account maintains a [threshold](/home/glossary#threshold), the minimum number of owners required to confirm a transaction before it gets executed. The threshold of owners required to confirm a transaction is also stored in the Safe storage. Owners of a Safe account can change the threshold value as well. A Safe account can have any number of owners, and the threshold value can be set between one and the total number of owners.

Reviewer: replace Safe with Safe Smart Account, or just say 'in the storage'.

#### Signature verification

> Safe account, being a contract account, does not have a private key to sign transactions, and EVM cannot verify incoming transactions to a contract account. Hence, a contract account has to do the authentication and authorization in its code. When a transaction is submitted to a Safe account, it is first verified by the Safe account to ensure that the transaction is valid. If the required number of owners has signed the transaction, the transaction is allowed to be executed. If the required number of owners has not signed the transaction, the transaction reverts to the signature validation step.

Reviewer: Safe Smart Account.

> A Safe account verifies if each signer is an owner of the Safe account and verifies the signature based on the signature type. To learn more about the signature types supported by Safe and encoding, refer to the [Signatures](/smart-account-signatures) page.

Reviewer: Safe Smart Account

#### Transaction flow

> Transactions through a Safe account can be primarily divided into two types:

Reviewer: Safe Smart Account

##### Safe Transaction

> Safe Smart Account contract provides `execTransaction` function to submit and execute a Safe transaction which is signed by the owners of the Safe account.

Reviewer: Safe Smart Account (at the end)

> To execute a transaction with the Safe, the `execTransaction` method needs to be called with the following parameters:

Reviewer: Safe Smart Account

import SafeSmartAccount from '../../assets/svg/ic-smart-account.svg'
import CustomCard from '../../components/CustomCard'

> ## EOAs

Reviewer 2: Having EOAs as the page title is a bit confusing for me. How about something like `How do Safe Smart Accounts work?` and the EOAs as H2?

Author: https://github.com/safe-global/safe-docs/pull/653/commits/43f9f5d563e5d25cb184f9eef1dd2f54ad10b99a

> EOAs for short are Ethereum accounts that use traditional key pairs. That is, they consist of a single private key that can be used to make transactions and sign messages. If you gain access to that private key, you gain full control of the account. This means that your private key is a single point of failure that stands between you and your funds being lost.

Reviewer 2: We should link to the glossary here: https://docs.safe.global/home/glossary#externally-owned-account

> ## Contract accounts

Reviewer 2: We should consider if we wanna call them contract accounts or smart accounts as we refer to them as smart account elsewhere.

> The other type of Ethereum accounts are smart contract accounts. Like EOAs, smart contract accounts each have a unique public Ethereum address, and it is impossible to tell them apart from EOAs by looking at an Ethereum address. Smart contract accounts too can receive funds and make transactions like EOAs. Generally, the key difference is that no single private key is used to verify transactions. Instead, the logic behind how the account completes transactions is defined in the smart contract code. Smart contracts are programs that run on the Ethereum blockchain and execute when specific conditions are met. Their functionality within contract accounts means that such accounts, in contrast to EOAs, can, for example, implement access rights that specify by whom, how, and under which conditions transactions can be executed, as well as more complex logic.

Reviewer 2: We should cross link to https://docs.safe.global/home/glossary#smart-account and aim to reduce redundancy, if possible.

> # EOAs vs. Contract Accounts

Reviewer 2: ```suggestion
# Difference between EOAs and Contract Accounts
```

Reviewer 2: Only the page title should be H1.

> What is the difference between the two different types of accounts on Ethereum: externally owned accounts (EOAs) and contract accounts?

Reviewer 2: ```suggestion
```

  </Grid>
</Grid>

### Features

#### High Security

> Safe's **multi-signature** functionality allows you to define a list of owner accounts and a threshold number of accounts required to confirm a transaction. Once the threshold of owner accounts have confirmed a transaction, the Safe transaction can be executed. Owners can either be EOAs, other smart contract accounts or even a Passkey.

Reviewer 2: ```suggestion
Safe's **multi-signature** functionality allows you to define a list of owner accounts and a threshold number of accounts required to confirm a transaction. Once the threshold of owner accounts have confirmed a transaction, the Safe transaction can be executed. Owners can either be EOAs, other smart contract accounts or even a passkey.
```

#### Advanced execution logic

> It is possible to make use of different **Safe Library contracts** to perform complex transactions. A very common example of this is **batched transactions** where multiple simple Ethereum transactions are combined and executed at once. That means instead of having to sign several transactions sequentially, a user just needs to sign one batched transaction.

Reviewer 2: A bunch of other terms here should be cross-linked to the Glossary.

> #### Advanced access management

Author: Recovery is not a core bulit-in feature of Safe Smart Accounts. Users can enable recovery mechanism using modules.

You can add **Safe Modules** to your Safe which facilitates more fine-grained access management. For instance, it is possible to define a module that can only be used to **recover access** to a Safe under specific circumstances. Another example is **allowance modules** that allow owners of a Safe to grant limited execution permission, such as a daily limit to external accounts.

#### Token callback support

> Many new tokens require wallet contracts to implement callbacks. Token standards like **ERC721** and **ERC1155** allow contracts to immediately react to receiving tokens through these and make it even possible to reject the transfer altogether.

Reviewer 2: ```suggestion
Many new tokens require wallet contracts to implement callbacks. Token standards like **ERC-721** and **ERC-1155** allow contracts to immediately react to receiving tokens through these and make it even possible to reject the transfer altogether.
```

Reviewer 2: We use this format throughout the docs. Could you please update elsewhere in the docs please?

##### Guard Management

Guards are used to check if a transaction should be executed or rejected based on the logic defined in the guard. A Guard Manager contract allows you to add, remove, and replace guards. Guards are security critical as a malicious guard could prevent transactions from being executed and block access to funds stored in the Safe. Events are emitted whenever a guard is updated.

#### SafeProxy

A Safe Proxy is a contract that delegates all calls to the Safe Singleton. By deploying a Proxy the cost to create a Safe account gets reduced as the byte code size of proxy contract is lesser than deploying the actual Safe contract code.

![Safe Proxy Creation](../../assets/diagram-safe-smart-account-proxy-creation.png)

### Concepts

#### Owners

Each Safe account maintains it own list of owners in the storage in the form of Ethereum addresses. Owners can be added or removed by other owners. Along with list of owners, a Safe account also maintains a **threshold** which is the minimum number of owners required to confirm a transaction before it gets executed.

> The threshold of owners required to confirm a transaction is also stored in the Safe storage. Owners of a Safe account can change the threshold value as well. A Safe account can have any number of owners and the threshold value can be set to any value between 1 and the total number of owners.

Reviewer 2: ```suggestion
The threshold of owners required to confirm a transaction is also stored in the Safe storage. Owners of a Safe account can change the threshold value as well. A Safe account can have any number of owners and the threshold value can be set to any value between one and the total number of owners.
```
